OWML.ModHelper.Menus/ModLayoutButton.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Reflection;
+using OWML.Common.Menus;
+using UnityEngine;
+using UnityEngine.UI;
+using UnityEngine.EventSystems;
+
+namespace OWML.ModHelper.Menus
+{
+    public class ModLayoutButton : IModLayoutButton
+    {
+        public event Action OnClick;
+
+        public Button Button { get; }
+        public IModMenu Menu { get; private set; }
+        public HorizontalLayoutGroup LayoutGroup { get; private set; }
+        private UIStyleApplier _buttonStyleApplier;
+
+        private int _index;
+        public int Index
+        {
+            get => Button.transform.parent == null ? _index : Button.transform.GetSiblingIndex();
+            set
+            {
+                _index = value;
+                Button.transform.SetSiblingIndex(value);
+            }
+        }
+        public ModLayoutButton(Button button, IModMenu menu)
+        {
+            Button = button;
+            Button.onClick.AddListener(() => OnClick?.Invoke());
+            var temp = Button.GetComponentInChildren<Text>().gameObject;
+            GameObject.Destroy(temp);
+            temp = new GameObject("LayoutGroup", new Type[] { typeof(RectTransform) });
+            temp.transform.SetParent(button.transform);
+            var img = temp.AddComponent<Image>();
+            img.raycastTarget = true;
+            img.color = new Color(255, 255, 255, 0);
+            LayoutGroup = temp.AddComponent<HorizontalLayoutGroup>();
+            Initialize(menu);
+            this._buttonStyleApplier = Button.GetComponent<UIStyleApplier>();
+            LayoutGroup.childControlWidth = false;
+            LayoutGroup.childControlHeight = false;
+            LayoutGroup.childForceExpandHeight = false;
+            LayoutGroup.childForceExpandWidth = false;
+            LayoutGroup.childAlignment = TextAnchor.MiddleCenter;
+            LayoutGroup.transform.localPosition = Vector3.zero;
+            ((RectTransform)LayoutGroup.transform).pivot = new Vector2(0.5f, 0.5f);
+            UpdateState();
+        }
+        public void UpdateState()
+        {
+            FieldInfo texts = typeof(UIStyleApplier).GetField("_textItems", BindingFlags.NonPublic | BindingFlags.Instance);
+            FieldInfo foregrounds = typeof(UIStyleApplier).GetField("_foregroundGraphics", BindingFlags.NonPublic | BindingFlags.Instance);    
```

Comment:
Do these two need to be fetched every time `UpdateState` is called? Can't we have them in a class attribute and set it up just once in `Initialize` or somewhere else?

Reply:
done